File: src/2015/day12/day12.py
import json
import sys

from libs.aoc_runner.python import aoc_runner


def solve(data, validate=False):
    if data == None:
        return 0

    if isinstance(data, str):
        if validate and data == "red":
            return 0, False
        return 0, True

    if isinstance(data, int):
        return data, True

    if isinstance(data, list):
        sum = 0
        for obj in data:
            value, valid = solve(obj, validate)
            sum += value
        return sum, True

    sum = 0
    for [key, val] in data.items():
        value, valid = solve(val, validate)
        if not valid:
            return 0, True
        sum += value
    return sum, True


def samples():
    pass


def solve_1(x):
    data = json.loads(x[0])
    res, _ = solve(data)
    print(f"part 1: {res}")
    assert res == 119433


def solve_2(x):
    data = json.loads(x[0])
    res, _ = solve(data, True)
    print(f"part 2: {res}")
    assert res == 68466


if __name__ == "__main__":
    with open("day12.input", "r", encoding="utf8") as f:
        inp = f.readlines()
    sys.exit(
        aoc_runner.aoc_runner(
            samples,
            solve_1,
            solve_2,
            inp,
        )
    )

File: src/2015/day12/day12.cpp
/**
 * https://adventofcode.com/2015/day/12
 */
#include <algorithm>
#include <cassert>
#include <icecream.hpp>
#include <numeric>
#include <regex>
#include <string>
#include <vector>

#include "aoc_runner.h"
#include "aoc_utils.h"
#include "string_utils.h"

int64_t solve_1(std::vector<std::string> const& inp) {
    auto numbers = string_utils::numbers_from_string(inp[0]);
    return std::accumulate(numbers.cbegin(), numbers.cend(), static_cast<int64_t>(0));
}

int64_t solve_2(std::vector<std::string> const& inp_) {
    // Moved over to python for this one.
    return 0;
}

void samples() {}

int main(int argc, char** argv) {
    auto input = aoc::utils::read_input(AOC_INPUT);

    auto solve_1_wrapper = [](std::vector<std::string> const& inp) -> void {
        auto part1 = solve_1(inp);
        std::cout << "part 1: " << part1 << std::endl;
    };

    auto solve_2_wrapper = [](std::vector<std::string> const& inp) -> void {
        auto part2 = solve_2(inp);
        assert(part2 > 43050);
        std::cout << "part 2: " << part2 << std::endl;
    };

    return aoc::run(argc, argv, samples, solve_1_wrapper, solve_2_wrapper, input);
}
